src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -468,10 +469,13 @@ protected override void Dispose(bool disposing)
                 _canRead = false;
                 _canWrite = false;
                 _canSeek = false;
-                // Don't set the buffer to null, to avoid a NullReferenceException
-                // when users have a race condition in their code (ie, they call
-                // Close when calling another method on Stream like Read).
-                //_buffer = null;
+
+                if (_buffer != null)
+                {
+                    ArrayPool<byte>.Shared.Return(_buffer, clearArray: true);
```

Comment:
This means that accidental races in user code - like what is mentioned in the now-deleted comment - can lead to very interesting data corruptions now. It may be the right trade-off, but I am wondering whether we should be doing steps to mitigate the potential damage:
-  Try to detect the race - e.g. if the safe handle dispose operation does not actually close the safe handle, it means that there is other thread using the safe handle and likely the buffer as well, and thus it is dangerous to return the array back to the pool
and/or
- Reduce the window where the race can cause data corruption - e.g. set the _buffer to null first, and then return it to the pool.